core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1057,7 +1057,8 @@ class Log(@volatile private var _dir: File,
                      leaderEpoch: Int,
                      origin: AppendOrigin = AppendOrigin.Client,
                      interBrokerProtocolVersion: ApiVersion = ApiVersion.latestVersion): LogAppendInfo = {
-    append(records, origin, interBrokerProtocolVersion, assignOffsets = true, leaderEpoch, ignoreRecordSize = false)
+    val validateAndAssignOffsets = origin != AppendOrigin.RaftLeader
+    append(records, origin, interBrokerProtocolVersion, validateAndAssignOffsets, leaderEpoch, ignoreRecordSize = false)
```

Comment:
It is a little odd here is that we have to pass through `leaderEpoch` even though we expect the Raft leader to have set it already. Perhaps we should be validating it in `analyzeAndValidateRecords`. We can verify for each batch that the leader epoch matches when the append origin is `RaftLeader`. Does that make sense?

Reply:
Make sense to me, it is also a little bit odd to me, but I put it here because I think `assignOffsets`==true for `appendAsLeader` and ==false for `appendAsFollower`, which means normally `assignOffsets` is determined by the caller, the `RaftLeader` is just a special case for `appendAsLeader`, if we move the logic in `analyzeAndValidateRecords`, that means it need to determine whether to `assignOffsets` without caller info, does that doable?

Reply:
The call to `analyzeAndValidateRecords` is done both for leader as well as follower appends. Basically we do a shallow iteration over the batches in order to collect some information and validate the CRC. My thought was to pass `leaderEpoch` into `analyzeAndValidateRecords` and add a basic check like this:
```java
records.batches.forEach { batch =>
  ...
  if (origin === RaftLeader && batch.partitionLeaderEpoch != leaderEpoch) {
    throw new InvalidRecordException("Append from Raft leader did not set the batch epoch correctly")
  }
}  
```